src/charm.py

```diff
@@ -4,27 +4,346 @@
+                    self.unit.status = BlockedStatus("Unable to create cluster")
+                    logger.debug("Unable to create cluster: {}".format(e))
+                    return
+            else:
+                # When unit is not the leader, it should wait
+                # for the leader to configure it a cluster node
+                self.unit.status = WaitingStatus("Waiting for instance to join the cluster")
+                # Create control file in data directory
+                container.push(CONFIGURED_FILE, make_dirs=True, source="configured")
+                return
+
+        else:
+            # Configure the layer when changed
+            current_layer = container.get_plan()
+            new_layer = self._pebble_layer
+
+            if new_layer.services != current_layer:
+                logger.info("Add pebble layer")
+                container.add_layer(MYSQLD_SERVICE, new_layer, combine=True)
+                container.restart(MYSQLD_SERVICE)
+                self._mysql._wait_until_mysql_connection()
 
-    def _on_mysql_pebble_ready(self, _):
-        """Define and start a workload using the Pebble API."""
         self.unit.status = ActiveStatus()
 
+    def _on_peer_relation_joined(self, event: RelationJoinedEvent):
+        """Handle the peer relation joined event."""
+        # Only leader unit add instances to the cluster
+        if not self.unit.is_leader():
+            return
+
+        # Defer run when leader is not active
+        if not isinstance(self.unit.status, ActiveStatus):
+            event.defer()
+            return
+
+        new_instance = self._get_unit_fqdn(event.unit.name)
+
+        # Check if new instance is ready to be added to the cluster
+        if not self._mysql.is_instance_configured_for_innodb(new_instance):
+            event.defer()
+            return
+
+        # Check if instance was already added to the cluster
+        if self._mysql.is_instance_in_cluster(self._get_unit_hostname(event.unit.name)):
+            logger.debug(f"Instance {new_instance} already in cluster")
+            return
+
+        # Add new instance to ipAllowlist global variable
+        peer_data = self._peers.data[self.app]
+        if new_instance not in peer_data.get("allowlist").split(","):
+            peer_data["allowlist"] = f"{peer_data['allowlist']},{new_instance}"
+            try:
+                self._mysql.update_allowlist(peer_data["allowlist"])
+            except MySQLUpdateAllowListError:
+                logger.debug("Unable to update allowlist")
+                event.defer()
+                return
+
+        # Add new instance to the cluster
+        try:
+            self._mysql.add_instance_to_cluster(new_instance)
+            logger.debug(f"Added instance {new_instance} to cluster")
+
+        except MySQLAddInstanceToClusterError:
+            logger.debug(f"Unable to add instance {new_instance} to cluster.")
+            event.defer()
```

Comment:
I'm not sure if we should block without retrying, due to some transient issues. But on the other hand it could be deferring the event forever. Not sure what's the best approach